index.js

```diff
@@ -70,10 +70,13 @@ function flameGraph (opts) {
   var hoverFrame = null
   var currentAnimation = null
 
+  // Use custom tooltip rendering function if defined
+  if (opts.renderTooltip !== undefined) renderTooltip = node => opts.renderTooltip && opts.renderTooltip(node)
+
   // Use custom coloring function if one has been passed in
-  if (opts.colorHash) colorHash = (d, decimalAdjust, allSamples, tiers) => {
-    return opts.colorHash(stackTop, { d, decimalAdjust, allSamples, tiers })
-  }
+  if (opts.colorHash !== undefined) colorHash = (d, decimalAdjust, allSamples, tiers) => 
+    opts.colorHash ? opts.colorHash(stackTop, { d, decimalAdjust, allSamples, tiers }) : frameColors.fill
```

Comment:
doesn't the global `colorHash` function get overridden only if opts.colorHash is defined?
Or am I missing something?

Reply:
Yeah. So it would leak between d3-fg instances if there are more than one.

Reply:
I see what you mean...
well spotted, I'll fix that.

Reply:
Hmm, good point. I didn't think about multiple instances of d3-fg when I did the colorHash thing a while back. 

Two things: 

1. The same will also be true of `renderTooltip`
2. There will be many other issues with potentially having multiple instances of d3-fg, potentially different versions, since most of the functions are global and (if it's just put in a script tag) there's no closure or block scoping. 

I think a better solution than adding another namespace `coloringFunction` and more complexity like below would be defining the default functions using `let colorHash = function (...` and `let renderTooltip = function (...` (maybe this is what you meant @goto-bus-stop by local variables?). 

That will scope to the block which will avoid the problem if there are multiple versions browserified etc (the most common case). Also, it's good for readability - it tells a reader these functions can be overridden.

Then we should put looking into the whole issue of multiple instances of d3-fg as a separate issue (much lower priority - unless there's a big use case)?

Reply:
renderTooltip is already defined in the closure, so it doesn't have this problem—overwriting it will only overwrite it locally.

Reply:
But yes, something like `let colorHash = opts.colorHash || defaultColorHash` and `let renderTooltip = opts.renderTooltip || defaultRenderTooltip` would be perfect, I think :)

Reply:
> if it's just put in a script tag

This isn't a concern, I think—d3-fg uses `require()` so you _have_ to bundle it first to use it in the browser. Almost all bundlers wrap their output in a closure, the only one I know of that doesn't is rollup but it doesn't do `require()`.

Reply:
I like the `default...` bit but the exact suggestion doesn't take passing `null` into account. And we don't need `let` in that scenario since it's not redefined. 

So I think what would work best is:

```
const colorHash = (opts.colorHash === undefined) ? defaultColorHash : coloringFunction = (...

const renderTooltip = (opts.renderTooltip === undefined) defaultRenderTooltip :  node => ...

...

function defaultRenderTooltip (...

... 

function defaultColorHash (...
```

Reply:
no prob, I just pushed an update.
I also linted the code a bit